molgenis-app-gavin2/src/main/java/org/molgenis/app/gavin/GavinServiceImpl.java

```diff
@@ -0,0 +1,213 @@
+    FileMeta filteredInput = createEmptyFile("filteredInput.vcf");
+    FileMeta discardedInput = createEmptyFile("discardedInput.txt");
+
+    GavinRun gavinRun = createGavinRun(part, filteredInput, discardedInput);
+    LOG.info("GavinRun created: '{}'", gavinRun.getId());
+
+    Multiset<LineType> parsedLineTypes = parseInputFile(inputFile, filteredInput, discardedInput);
+    if (parsedLineTypes.count(LineType.VCF) == 0) {
+      fail(gavinRun.getId(), "No usable lines were found in the uploaded file");
+    }
+
+    dataService.delete(FILE_META, inputFile);
+
+    return gavinRun.getId();
+  }
+
+  private Multiset<LineType> parseInputFile(
+      FileMeta inputFile, FileMeta filteredInput, FileMeta discardedInput) throws IOException {
+    Multiset<LineType> parsedLineTypes =
+        parser.tryTransform(
+            fileStore.getFile(inputFile.getId()),
+            fileStore.getFile(filteredInput.getId()),
+            fileStore.getFile(discardedInput.getId()));
+
+    filteredInput.setSize(fileStore.getFile(filteredInput.getId()).length());
+    discardedInput.setSize(fileStore.getFile(discardedInput.getId()).length());
+    dataService.update(FILE_META, Stream.of(filteredInput, discardedInput));
+    return parsedLineTypes;
+  }
+
+  private GavinRun createGavinRun(Part part, FileMeta filteredInput, FileMeta discardedInput) {
+    GavinRun gavinRun = gavinRunFactory.create();
+    gavinRun.setId(idGenerator.generateId(Strategy.LONG_SECURE_RANDOM));
+    gavinRun.setInputFileName(part.getSubmittedFileName());
+    gavinRun.setFilteredInputFile(filteredInput);
+    gavinRun.setDiscardedInputFile(discardedInput);
+    gavinRun.setSubmittedAt(Instant.now());
+    gavinRun.setStatus(Status.PENDING);
+    dataService.add(GAVIN_RUN, gavinRun);
+    return gavinRun;
+  }
+
+  @Override
+  public GavinRun get(String id) {
+    LOG.debug("Getting GavinRun '{}'", id);
+
+    GavinRun gavinRun = dataService.findOneById(GAVIN_RUN, id, GavinRun.class);
+    if (gavinRun == null) {
+      throw new UnknownEntityException(GAVIN_RUN, id);
+    }
+    return gavinRun;
+  }
+
+  @Override
+  @Transactional
+  public void start(String id) {
+    LOG.info("GavinRun has started: '{}'", id);
+
+    GavinRun gavinRun = get(id);
+    gavinRun.setStartedAt(Instant.now());
+    gavinRun.setStatus(Status.RUNNING);
+    dataService.update(GAVIN_RUN, gavinRun);
+  }
+
+  @Override
+  @Transactional
+  public void finish(String id, String log, HttpServletRequest httpServletRequest)
+      throws IOException {
+    LOG.info("GavinRun has finished: '{}'", id);
+
+    GavinRun gavinRun = get(id);
+    FileMeta outputFile;
+    try {
+      outputFile = storeUploadedFile(httpServletRequest.getPart("outputFile"));
+    } catch (ServletException e) {
+      throw new IllegalStateException("Request is not of type multipart/form-data");
+    }
+
+    gavinRun.setOutputFile(outputFile);
+    gavinRun.setLog(gavinRun.getLog() + log);
```

Comment:
Persistence will fail when your log length is larger than the allowed maximum length resulting in jobs that never finish (and resources that are never removed --> see suggestion on cleaning up resources for jobs that have been alive for > maximum allowed lifetime)